main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1715 @@
+    * Note: We use `subst.extended` to check for conflicts. For example, if we already know that `s = [x -> c17]` and we
+    * learn that `x -> true` then we will try to extend s with the new binding which will raise a [[ConflictException]].
+    */
+  private def propagateUnit(l: List[Equation], s: BoolSubstitution): (List[Equation], BoolSubstitution) = {
+    var pending = l
+    var subst = s
+
+    // We iterate until no changes are detected.
+    var changed = true
+    while (changed) {
+      changed = false
+
+      var rest: List[Equation] = Nil
+      for (e <- pending) {
+        e match {
+          // Case 1: x ~ true
+          case Equation(Term.Var(x), Term.True, loc) =>
+            subst = subst.extended(x, Term.True, loc)
+            changed = true
+
+          // Case 2: x ~ c
+          case Equation(Term.Var(x), Term.Cst(c), loc) =>
+            subst = subst.extended(x, Term.Cst(c), loc)
+            changed = true
+
+          // Case 3: x /\ y /\ z /\... ~ true
+          case Equation(Term.And(csts, vars, rest), Term.True, loc) if csts.isEmpty && rest.isEmpty =>
+            for (Term.Var(x) <- vars) {
+              subst = subst.extended(x, Term.True, loc)
+              changed = true
+            }
+
+          case _ =>
+            rest = e :: rest
+        }
+      }
+      // INVARIANT: We apply the current substitution to all remaining equations.
+      pending = subst(rest)
+    }
+
+    // Reverse the unsolved equations to ensure they are returned in the original order.
+    (pending.reverse, subst)
+  }
+
+  /**
+    * Propagates variables through the equation system. Cannot fail.
+    *
+    * The observation is that we if we have simple equations of the form: `x ~ y`
+    * then we can create a substitution that binds `[x -> y]`. Every time we create
+    * a binding, we must be careful to update both unsolved and pending equations.
+    *
+    * For example, if the equation system is:
+    *
+    * {{{
+    *     x78914 ~ (c1498 ∧ c1500 ∧ c1501)
+    *     x78914 ~ (x78926 ∧ x78923 ∧ x78917)
+    *     x78917 ~ x127244
+    *     x78921 ~ x127251
+    *     x78923 ~ (x127249 ∧ x127247 ∧ x127248)
+    *     x78926 ~ (x127254 ∧ x127252)
+    * }}}
+    *
+    * then after variable propagation it is:
+    *
+    * {{{
+    *     x78914 ~ (c1498 ∧ c1500 ∧ c1501)
+    *     x78914 ~ (x78926 ∧ x78923 ∧ x127244)
+    *     x78923 ~ (x127249 ∧ x127247 ∧ x127248)
+    *     x78926 ~ (x127254 ∧ x127252)
+    * }}}
+    *
+    * with the substitution:
+    *
+    * {{{
+    *      x78917 -> x127244
+    *     x127251 -> x78921
+    * }}}
+    *
+    * Returns a list of unsolved equations and a partial substitution.
+    */
+  private def propagateVars(l: List[Equation], s: BoolSubstitution): (List[Equation], BoolSubstitution) = {
+    var pending = l
+    var subst = s
+
+    var unsolved: List[Equation] = Nil
+
+    // INVARIANT: The current substitution has been applied to BOTH unsolved AND pending equations.
```

Comment:
Is it not better to keep a running substitution instead of this n^2 substitution application?

Reply:
What do you mean?

Reply:
Note that all functions other than SVE are very fast because they operate on typically less than 50 constraints.